File: handin/lab5/lab5.py
#!/usr/bin/env python
#encoding: utf-8

import sys


#
# xrange() is replaced by range() in Python 3
#
if (sys.hexversion >= 0x300000):
    def xrange(stop):
        return range(stop)



def main():
    """
    Main function
    """

    # Check if we have proper arguments
    if (len(sys.argv) - 1 < 2):
        sys.stderr.write("format: %s <in-file> <out-file>\n" % sys.argv[0])
        exit(1)

    # Open input and output files
    try:
        fin = open(sys.argv[1], "r")
    except IOError as e:
        sys.stderr.write("can't open `%s': %s\n" % (sys.argv[1], e.strerror))
        exit(2)
    try:
        fout = open(sys.argv[2], "w")
    except IOError as e:
        sys.stderr.write("can't open `%s': %s\n" % (sys.argv[2], e.strerror))
        exit(2)

    # Process the input file
    records = {}
    maxlengths = {}
    for line in fin:
        # split 'line' into fields
        fields = line.rstrip('\n').split(',')
        n_arr = []
        v_arr = []
        rec = {}
        for f in fields:
            # split each field into 'name' and 'value'
            assign = f.split('=')
            name = assign[0].strip(' ')
            value = assign[1].strip(' ')
            n_arr.append(name)
            v_arr.append(value)
            
            # update the maximum length
            if (maxlengths.has_key(name)):
                if (len(name) > maxlengths[name]):
                    maxlengths[name] = len(name)
                if (len(value) > maxlengths[name]):
                    maxlengths[name] = len(value)
            
            # check if found the name assignment
            if (name == "name"):
                line_name = value
            else:
                rec[name] = value

        # update the dictionary
        records[line_name] = rec

    # Close the files
    fout.close()
    fin.close()


#
# Call main() if the program is not being used as a module
#
if (__name__ == "__main__"):
    main()

File: handin/lab5/lab5.pl
#!/usr/bin/env perl
#

# Check if we have proper arguments
die "format: %0 <in-file> <out-file>\n" if ($#ARGV + 1 < 2);

# Open input and output files
open(my $fin, "<", $ARGV[0]) ||
    die "$0: can't open \`$ARGV[0]\': $!\n";
open(my $fout, ">", $ARGV[1]) ||
    die "$0: can't open \`$ARGV[1]\': $!\n";

# Process the input file
my %records;
my %maxlengths;
my $line_name;
while (<$fin>) {
    my %rec;
    chomp;
    @fields = split /,/;
    foreach $f (@fields) {
	# split each field into 'name' and 'value'
	@assign = split /=/, $f;
	$name = $assign[0];
	$name =~ s/^\s+|\s+$//g;
	$value = $assign[1];
	$value =~ s/^\s+|\s+$//g;

	# update the maximum length
	if ((!exists $maxlengths{$name}) || (length($name) > $maxlengths{$name})) {
	    $maxlengths{$name} = length($name);
	}
	if (length($value) > $maxlengths{$name}) {
	    $maxlengths{$name} = length($value);
	}

	# check if found the name assignment
	if ($name eq "name") {
	    $line_name = $value;
	} else {
	    $rec{$name} = $value;
	}
    }

    # update the dictionary
    if (exists $records{$line_name}) {
	die "name \`$line_name\' occurs more than once\n";
    }

    $records{$line_name} = %rec;
} # while (<$fin>)

# Close the input file
close($fin) || 
    die "$0: can't close \`$ARGV[0]\': $!\n";

# Sort the field names
@sorted_fields = sort(keys(%maxlengths));

# Print the header
$line = sprintf("%" . $maxlengths{"name"} . "s  ", "name");

foreach $n (@sorted_fields) {
    if (($n ne "name") && ($n ne "address")) {
	$line .= sprintf("%" . $maxlengths{$n} . "s  ", $n);
    }
}
$line .= sprintf("%" . $maxlengths{"address"} . "s\n", "address");
print $line;

# Sort the dictionary
@sorted_records = sort(keys(%records));
foreach $name (@sorted_records) {
    $line = sprintf("%" . $maxlengths{"name"} . "s  ", $name);
    foreach $f (@sorted_fields) {
	if (($f ne "name") && ($f ne "address")) {
	    print "checking record $f for name $name: $records{$name}{$f}\n";
	    $line .= sprintf("%" . $maxlengths{$f} . "s  ",
			     (exists $records{$name}{$f}) ?
			      $records{$name}{$f} : "");
	}
    }

    # address may not be present
    $line .= sprintf("%" . $maxlengths{"address"} . "s\n",
		     (exists $records{$name}{"address"}) ?
		     $records{$name}{"address"} : "");

    print $line;
}

# Close the output file
close($fout) || 
    die "$0: can't close \`$ARGV[1]\': $!\n";
